color/src/color.rs

```diff
@@ -0,0 +1,531 @@
+        let cs = PhantomData;
+        Self { components, cs }
+    }
+
+    #[must_use]
+    pub fn convert<U: Colorspace>(self) -> PremulColor<U> {
+        if TypeId::of::<T>() == TypeId::of::<U>() {
+            PremulColor::new(self.components)
+        } else if U::IS_LINEAR && T::IS_LINEAR {
+            let (multiplied, alpha) = split_alpha(self.components);
+            let lin_rgb = T::to_linear_srgb(multiplied);
+            PremulColor::new(add_alpha(U::from_linear_srgb(lin_rgb), alpha))
+        } else {
+            self.un_premultiply().convert().premultiply()
+        }
+    }
+
+    #[must_use]
+    pub fn un_premultiply(self) -> AlphaColor<T> {
+        let (multiplied, alpha) = split_alpha(self.components);
+        let scale = if alpha == 0.0 { 1.0 } else { 1.0 / alpha };
+        AlphaColor::new(add_alpha(T::LAYOUT.scale(multiplied, scale), alpha))
+    }
+
+    /// Interpolate colors.
+    ///
+    /// Note: this function doesn't fix up hue in cylindrical spaces. It is
+    /// still be useful if the hue angles are compatible, particularly if
+    /// the fixup has been applied.
+    #[must_use]
+    pub fn lerp_rect(self, other: Self, t: f32) -> Self {
+        self + t * (other - self)
+    }
+
+    /// Apply hue fixup for interpolation.
+    ///
+    /// Adjust the hue angle of `other` so that linear interpolation results in
+    /// the expected hue direction.
+    pub fn fixup_hues(self, other: &mut Self, direction: HueDirection) {
+        if let Some(ix) = T::LAYOUT.hue_channel() {
+            fixup_hue(self.components[ix], &mut other.components[ix], direction);
+        }
+    }
+
+    /// Linearly interpolate colors, with hue fixup if needed.
+    #[must_use]
+    pub fn lerp(self, mut other: Self, t: f32, direction: HueDirection) -> Self {
+        self.fixup_hues(&mut other, direction);
+        self.lerp_rect(other, t)
+    }
+
+    #[must_use]
+    pub const fn mul_alpha(self, rhs: f32) -> Self {
+        let (multiplied, alpha) = split_alpha(self.components);
+        Self::new(add_alpha(T::LAYOUT.scale(multiplied, rhs), alpha * rhs))
+    }
+
+    /// Difference between two colors by Euclidean metric.
+    #[must_use]
+    pub fn difference(self, other: Self) -> f32 {
+        let x = self.components;
+        let y = other.components;
+        let (d0, d1, d2, d3) = (x[0] - y[0], x[1] - y[1], x[2] - y[2], x[3] - y[3]);
+        (d0 * d0 + d1 * d1 + d2 * d2 + d3 * d3).sqrt()
+    }
+}
+
+// Lossless conversion traits.
+
+impl<T: Colorspace> From<OpaqueColor<T>> for AlphaColor<T> {
+    fn from(value: OpaqueColor<T>) -> Self {
+        value.with_alpha(1.0)
+    }
+}
+
+impl<T: Colorspace> From<OpaqueColor<T>> for PremulColor<T> {
+    fn from(value: OpaqueColor<T>) -> Self {
+        Self::new(add_alpha(value.components, 1.0))
+    }
+}
+
+// Arithmetic traits. A major motivation for providing these is to enable
+// weighted sums, which are well defined when the weights sum to 1. In
```

Comment:
I don't have the domain knowledge here, but this sentence *suggests* that the operations implemented here are potentially suspect, and might be better as methods with documentation and caveats.

At the very least, the impls themselves probably should have documentation explaining some of this.

Reply:
I've thought about this deeply and feel that arithmetic operations are worthwhile to include. Basically, it enables spline interpolation with a generic spline implementation that has the arithmetic traits as bounds, and to me that upside outweighs any downside. Other color crates impl these as well.

I have moved this comment to documentation.

Reply:
Seems reasonable. An alternative curve bending design might be to have an `Arithmetic<T>` wrapper, where you're opting in to thinking carefully about this.